pkg/virt-handler/device-manager/mediated_device.go

```diff
@@ -0,0 +1,437 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package device_manager
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/util"
+	pluginapi "kubevirt.io/kubevirt/pkg/virt-handler/device-manager/deviceplugin/v1beta1"
+)
+
+const (
+	MDEV_RESOURCE_PREFIX = "MDEV_PCI_RESOURCE"
+)
+
+// Not a const for static test purposes
+var mdevBasePath string = "/sys/bus/mdev/devices"
+
+type MDEV struct {
+	UUID             string
+	typeName         string
+	parentPciAddress string
+	iommuGroup       string
+	numaNode         int
+}
+
+type MediatedDevicePlugin struct {
+	devs           []*pluginapi.Device
+	server         *grpc.Server
+	socketPath     string
+	stop           chan struct{}
+	health         chan string
+	devicePath     string
+	deviceName     string
+	resourceName   string
+	done           chan struct{}
+	deviceRoot     string
+	healthy        chan string
+	unhealthy      chan string
+	iommuToMDEVMap map[string]string
+	initialized    bool
+	lock           *sync.Mutex
+}
+
+func NewMediatedDevicePlugin(mdevs []*MDEV, resourceName string) *MediatedDevicePlugin {
+	s := strings.Split(resourceName, "/")
+	mdevTypeName := s[1]
+	serverSock := SocketPath(mdevTypeName)
+	iommuToMDEVMap := make(map[string]string)
+
+	initHandler()
+
+	devs := constructDPIdevicesFromMdev(mdevs, iommuToMDEVMap)
+	dpi := &MediatedDevicePlugin{
+		devs:           devs,
+		socketPath:     serverSock,
+		health:         make(chan string),
+		deviceName:     resourceName,
+		resourceName:   resourceName,
+		devicePath:     vfioDevicePath,
+		deviceRoot:     util.HostRootMount,
+		healthy:        make(chan string),
+		unhealthy:      make(chan string),
+		iommuToMDEVMap: iommuToMDEVMap,
+		initialized:    false,
+		lock:           &sync.Mutex{},
+	}
+
+	return dpi
+}
+
+func constructDPIdevicesFromMdev(mdevs []*MDEV, iommuToMDEVMap map[string]string) (devs []*pluginapi.Device) {
+	for _, mdev := range mdevs {
+		iommuToMDEVMap[mdev.iommuGroup] = mdev.UUID
+		dpiDev := &pluginapi.Device{
+			ID:     string(mdev.iommuGroup),
+			Health: pluginapi.Healthy,
+		}
+		if mdev.numaNode > 0 {
```

Comment:
maybe this is a dumb question. I just want to be sure this isn't meant to be >=0. Are we guaranteed there isn't a `0` numaNode?

Reply:
I actually overlooked this :( Of course, it should be =>0.